File: src/server/graphql/inbox.rs
//! The per-teammate email read: `Company.inboxes` over the [`InboxStore`] port.

use std::sync::Arc;

use async_graphql::{Context, ID, Object, SimpleObject};

use super::pagination::Page;
use crate::company::runtime::CompanyRuntime;
use crate::ports::inbox::{EmailRecord, InboxMeta};

/// One teammate inbox handle: metadata plus a paginated message resolver.
pub struct InboxGql {
    runtime: Arc<CompanyRuntime>,
    meta: InboxMeta,
    unread: i32,
}

impl InboxGql {
    fn new(runtime: Arc<CompanyRuntime>, meta: InboxMeta, unread: i32) -> Self {
        Self {
            runtime,
            meta,
            unread,
        }
    }
}

#[Object(name = "Inbox")]
impl InboxGql {
    /// The teammate local part / slug.
    async fn key(&self) -> ID {
        ID(self.meta.key.clone())
    }

    /// The inbox display name.
    async fn name(&self) -> String {
        self.meta.name.clone()
    }

    /// The full address (`{key}@{domain}` when a domain is configured).
    async fn address(&self) -> String {
        self.meta.address.clone()
    }

    /// The number of unread received messages.
    async fn unread(&self) -> i32 {
        self.unread
    }

    /// A page of this inbox's messages, oldest first.
    async fn messages(
        &self,
        #[graphql(default = 50)] first: i32,
        #[graphql(default = 0)] offset: i32,
    ) -> async_graphql::Result<Page<EmailMessageGql>> {
        let all = self
            .runtime
            .inbox()
            .messages(self.runtime.id(), &self.meta.key, usize::MAX, 0)
            .await?;
        let total = all.len() as i32;
        let items = all
            .into_iter()
            .skip(offset.max(0) as usize)
            .take(first.max(0) as usize)
            .map(EmailMessageGql::from)
            .collect();
        Ok(Page { items, total })
    }
}

/// One email in a teammate inbox. Mirrors [`EmailRecord`].
#[derive(SimpleObject)]
#[graphql(name = "EmailMessage")]
pub struct EmailMessageGql {
    /// The message id.
    pub id: ID,
    /// The sender's display name (may be empty).
    pub from_name: String,
    /// The sender's email address.
    pub from_email: String,
    /// The subject line.
    pub subject: String,
    /// The plain-text body.
    pub body: String,
    /// When it arrived / was sent, epoch millis.
    pub at_millis: f64,
    /// Whether the message has been read.
    pub read: bool,
    /// True for a sent message, false for a received one.
    pub outbound: bool,
}

impl From<EmailRecord> for EmailMessageGql {
    fn from(record: EmailRecord) -> Self {
        Self {
            id: ID(record.id),
            from_name: record.from_name,
            from_email: record.from_email,
            subject: record.subject,
            body: record.body,
            at_millis: record.at_millis as f64,
            read: record.read,
            outbound: record.outbound,
        }
    }
}

/// Resolves `Company.inboxes`, computing each inbox's unread count.
pub(crate) async fn resolve(
    ctx: &Context<'_>,
    runtime: &Arc<CompanyRuntime>,
) -> async_graphql::Result<Vec<InboxGql>> {
    let _ = ctx;
    let metas = runtime.inbox().inboxes(runtime.id()).await?;
    let mut out = Vec::with_capacity(metas.len());
    for meta in metas {
        let messages = runtime
            .inbox()
            .messages(runtime.id(), &meta.key, usize::MAX, 0)
            .await?;
        let unread = messages.iter().filter(|m| !m.outbound && !m.read).count() as i32;
        out.push(InboxGql::new(runtime.clone(), meta, unread));
    }
    Ok(out)
}

File: frontend/src/api/inbox.ts
// The live per-agent inbox API: the console reads each teammate's *real*
// correspondence through the host's `…/inboxes` routes (REST, camelCase over the
// wire) — the same `InboxStore` the inbound ingest webhook and every outbound
// send append to.
//
// This replaces the client-side `lib/inbox` localStorage fixture, which
// fabricated the same four emails ("Priya Sharma", "Stripe", "Weekly Digest",
// "Figma") for every teammate and could never show mail that actually arrived
// (issue #173). There is deliberately no seed and no local fallback: an empty
// inbox renders as empty, so a backend failure can't be masked by fake data.

import type { OpenCompanyClient } from "./client";

/** One email in a teammate's inbox — inbound or outbound. Mirrors `EmailRecord`. */
export interface EmailMessage {
  id: string;
  /** The inbox key this message is filed under (the teammate's agent id). */
  inbox: string;
  /** The sender's display name; may be empty (ingest doesn't always supply one). */
  fromName: string;
  fromEmail: string;
  subject: string;
  /** The plain-text body. There is no server-side preview; see `preview()`. */
  body: string;
  /** When it arrived / was sent, epoch millis. */
  atMillis: number;
  read: boolean;
  /** True for mail the company sent, false for mail it received. */
  outbound: boolean;
}

/** One teammate inbox as `GET …/inboxes` lists it. */
export interface Inbox {
  /** The inbox key — the teammate's agent id / address local part. */
  key: string;
  name: string;
  /** The full address (`{key}@{domain}`); empty until a domain is configured. */
  address: string;
  /** Whether it is receiving mail (the Team page toggle). */
  enabled: boolean;
  /** Unread *received* mail; a sent copy never counts. */
  unread: number;
  /** Every message filed here, inbound and outbound. */
  total: number;
}

/** A page of one inbox's mail, newest first. */
export interface MessagesPage {
  items: EmailMessage[];
  /** The unpaginated message count. */
  total: number;
}

/** The `PUT …/team/{agentId}/inbox` acknowledgement. */
export interface InboxAck {
  key: string;
  address: string;
}

/** Every inbox this company owns, enabled or not, with its unread count. */
export function listInboxes(
  client: OpenCompanyClient,
  company: string | null,
): Promise<Inbox[]> {
  return client.get<Inbox[]>(`${client.scopeFor(company)}/inboxes`);
}

/** One teammate's mail, newest first. */
export function inboxMessages(
  client: OpenCompanyClient,
  company: string | null,
  key: string,
  opts?: { limit?: number; offset?: number },
): Promise<MessagesPage> {
  const params = new URLSearchParams();
  if (opts?.limit !== undefined) params.set("limit", String(opts.limit));
  if (opts?.offset !== undefined) params.set("offset", String(opts.offset));
  const qs = params.toString();
  return client.get<MessagesPage>(
    `${client.scopeFor(company)}/inboxes/${encodeURIComponent(key)}/messages${qs ? `?${qs}` : ""}`,
  );
}

/**
 * Mark mail read — the given `ids`, or the whole inbox when omitted. Returns the
 * count still unread, so the badge follows the host rather than local state.
 */
export function markInboxRead(
  client: OpenCompanyClient,
  company: string | null,
  key: string,
  ids?: string[],
): Promise<{ unread: number }> {
  return client.post<{ unread: number }>(
    `${client.scopeFor(company)}/inboxes/${encodeURIComponent(key)}/read`,
    ids ? { ids } : {},
  );
}

/** Give a teammate an inbox, or take it away. Keyed by the roster **agent id**. */
export function setInboxEnabled(
  client: OpenCompanyClient,
  company: string | null,
  agentId: string,
  enabled: boolean,
): Promise<InboxAck> {
  return client.put<InboxAck>(
    `${client.scopeFor(company)}/team/${encodeURIComponent(agentId)}/inbox`,
    { enabled },
  );
}

/** Inboxes the operator has switched on, name-sorted for the selector. */
export function enabledInboxes(inboxes: Inbox[]): Inbox[] {
  return inboxes.filter((i) => i.enabled).sort((a, b) => a.name.localeCompare(b.name));
}

/**
 * A one-line preview of a message body. The host stores the full body only —
 * the fixture's hand-written `preview` field had no real counterpart — so the
 * list derives its snippet here.
 */
export function preview(body: string, max = 110): string {
  const flat = body.replace(/\s+/g, " ").trim();
  return flat.length > max ? `${flat.slice(0, max - 1).trimEnd()}…` : flat;
}
